File: binders/difido-testng/src/main/java/il/co/topq/difido/reporters/DifidoClient.java
package il.co.topq.difido.reporters;

import il.co.topq.difido.model.execution.MachineNode;
import il.co.topq.difido.model.remote.ExecutionDetails;
import il.co.topq.difido.model.test.TestDetails;

import java.io.File;

import org.apache.commons.httpclient.Header;
import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpMethod;
import org.apache.commons.httpclient.methods.PostMethod;
import org.apache.commons.httpclient.methods.PutMethod;
import org.apache.commons.httpclient.methods.RequestEntity;
import org.apache.commons.httpclient.methods.StringRequestEntity;
import org.apache.commons.httpclient.methods.multipart.FilePart;
import org.apache.commons.httpclient.methods.multipart.MultipartRequestEntity;
import org.apache.commons.httpclient.methods.multipart.Part;

import com.fasterxml.jackson.databind.ObjectMapper;

public class DifidoClient {

	private static final String BASE_URI_TEMPLATE = "http://%s:%d/api/";
	private final String baseUri;
	private final HttpClient client;

	public DifidoClient(String host, int port) {
		baseUri = String.format(BASE_URI_TEMPLATE, host, port);
		client = new HttpClient();
	}

	public int addExecution(ExecutionDetails details) throws Exception {
		final PostMethod method = new PostMethod(baseUri + "executions/");
		method.setRequestHeader(new Header("Content-Type", "application/json"));
		if (details != null) {
			final String descriptionJson = new ObjectMapper().writeValueAsString(details);
			method.setRequestEntity(new StringRequestEntity(descriptionJson,"application/json","UTF-8"));
		}
		final int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
		return Integer.parseInt(method.getResponseBodyAsString());
	}

	public void endExecution(int executionId) throws Exception {
		final PutMethod method = new PutMethod(baseUri + "executions/" + executionId + "?active=false");
		method.setRequestHeader(new Header("Content-Type", "text/plain"));
		final int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
	}

	public int addMachine(int executionId, MachineNode machine) throws Exception {
		PostMethod method = new PostMethod(baseUri + "executions/" + executionId + "/machines/");
		method.setRequestHeader(new Header("Content-Type", "application/json"));
		final ObjectMapper mapper = new ObjectMapper();
		final String json = mapper.writeValueAsString(machine);
		final RequestEntity entity = new StringRequestEntity(json,"application/json","UTF-8");
		method.setRequestEntity(entity);
		int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
		return Integer.parseInt(method.getResponseBodyAsString());
	}

	public void updateMachine(int executionId, int machineId, MachineNode machine) throws Exception {
		PutMethod method = new PutMethod(baseUri + "executions/" + executionId + "/machines/" + machineId);
		method.setRequestHeader(new Header("Content-Type", "application/json"));
		final ObjectMapper mapper = new ObjectMapper();
		final String json = mapper.writeValueAsString(machine);
		final RequestEntity entity = new StringRequestEntity(json,"application/json","UTF-8");
		method.setRequestEntity(entity);
		int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
	}

	public void addTestDetails(int executionId, TestDetails testDetails) throws Exception {
		PostMethod method = new PostMethod(baseUri + "executions/" + executionId + "/details");
		method.setRequestHeader(new Header("Content-Type", "application/json"));
		final ObjectMapper mapper = new ObjectMapper();
		final String json = mapper.writeValueAsString(testDetails);
		final RequestEntity entity = new StringRequestEntity(json,"application/json","UTF-8");
		method.setRequestEntity(entity);
		final int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
	}

	public void addFile(final int executionId, final String uid, final File file) throws Exception {
		PostMethod method = new PostMethod(baseUri + "executions/" + executionId + "/details/" + uid + "/file/");
		Part[] parts = new Part[] { new FilePart("file", file) };
		method.setRequestEntity(new MultipartRequestEntity(parts, method.getParams()));
		final int responseCode = client.executeMethod(method);
		handleResponseCode(method, responseCode);
	}

	private void handleResponseCode(HttpMethod method, int responseCode) throws Exception {
		if (responseCode != 200 && responseCode != 204) {
			throw new Exception("Request was not successful. Response is: " + responseCode + ".\n Response body: "
					+ method.getResponseBodyAsString());
		}

	}

}

File: binders/difido-dotnet/difido-client/Main/Report/Reporters/HtmlTestReporter/DifidoClient.cs
﻿using difido_client.Main.Report.Reporters.HtmlTestReporter.Model.Execution;
using difido_client.Report.Html.Model;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Web.Script.Serialization;

namespace difido_client.Main.Report.Reporters.HtmlTestReporter
{
    internal class DifidoClient
    {
        private readonly string BASE_URI_TEMPLATE = "http://{0}:{1}/api/";
        private readonly string baseUri;


        public DifidoClient(String host, int port)
        {
            baseUri = String.Format(BASE_URI_TEMPLATE, host, port);
        }

        public int AddExecution(ExecutionDetails details)
        {
            if (null == details)
            {
                throw new Exception("Execution details can't be null");
            }
            WebRequest request = WebRequest.Create(baseUri + "executions/");
            request.Method = "POST";         
            return Int32.Parse(SendContent(request, details));
        }

        public string endExecution(int executionId)
        {
            WebRequest request = WebRequest.Create(baseUri + "executions/"+executionId+"?active=false");
            request.Method = "PUT";
            request.ContentType = "text/plain";
            return Send(request);
         
        }


        public int AddMachine(int executionId, Machine machine)
        {
            WebRequest request = WebRequest.Create(baseUri + "executions/" + executionId + "/machines/");
            request.Method = "POST";
            return Int32.Parse(SendContent(request, machine));
        }

        private string SendContent(WebRequest request, Object obj)
        {

            request.ContentType = "application/json";
            string data = new JavaScriptSerializer().Serialize(obj);       
            byte[] byteArray = Encoding.UTF8.GetBytes(data);
            request.ContentLength = byteArray.Length;
            using (Stream dataStream = request.GetRequestStream())
            {                
                dataStream.Write(byteArray, 0, byteArray.Length);

            }            
            return Send(request);

        }

        public void UpdateMachine(int executionId, int machineId, Machine machine)
        {
            WebRequest request = WebRequest.Create(baseUri + "executions/" + executionId + "/machines/" + machineId);
            request.Method = "PUT";
            SendContent(request, machine);
        }

        public void AddTestDetails(int executionId, TestDetails testDetails)
        {
            WebRequest request = WebRequest.Create(baseUri + "executions/" + executionId + "/details");
            request.Method = "POST";
            SendContent(request, testDetails);
        }

        public void AddFile(int executionId, string uid, string file)
        {

            using (WebClient client = new WebClient())
            {
                client.UploadFile(baseUri + "executions/" + executionId + "/details/" + uid + "/file/", file);
            }
        }

        private string Send(WebRequest request)
        {
            HttpWebResponse response = null;
            string content = null;
            try
            {

                response = (HttpWebResponse)request.GetResponse();
                if (response.StatusCode != HttpStatusCode.OK && response.StatusCode != HttpStatusCode.Accepted && response.StatusCode != HttpStatusCode.NoContent)
                {
                    throw new Exception("Request was not successful. Response status is: " + response.StatusCode);
                }
                Stream dataStream = response.GetResponseStream();
                // Open the stream using a StreamReader for easy access.
                using (StreamReader reader = new StreamReader(dataStream))
                {
                    // Read the content.
                    content = reader.ReadToEnd();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
            finally
            {
                response.Close();
            }
            return content;
        }


    }
}
